src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -4399,48 +4399,94 @@ protected int notValidUtf8(Object string, int byteIndex,
     }
 
     @Primitive(name = "string_character_index", lowerFixnum = 2)
-    public abstract static class StringCharacterIndexPrimitiveNode extends PrimitiveArrayArgumentsNode {
+    @NodeChild(value = "string", type = RubyNode.class)
+    @NodeChild(value = "pattern", type = RubyNode.class)
+    @NodeChild(value = "offset", type = RubyNode.class)
+    public abstract static class StringCharacterIndexPrimitiveNode extends PrimitiveNode {
 
-        @TruffleBoundary
-        @Specialization
-        protected Object stringCharacterIndex(Object string, Object pattern, int offset,
-                @CachedLibrary(limit = "2") RubyStringLibrary stringLibrary,
-                @CachedLibrary(limit = "2") RubyStringLibrary patternLibrary,
-                @Cached RopeNodes.CalculateCharacterLengthNode calculateCharacterLengthNode) {
-            if (offset < 0) {
-                return nil;
-            }
+        @Child SingleByteOptimizableNode singleByteOptimizableNode = SingleByteOptimizableNode.create();
 
-            final Rope stringRope = stringLibrary.getRope(string);
-            final Rope patternRope = patternLibrary.getRope(pattern);
+        @CreateCast("string")
+        protected RubyNode coerceStringToRope(RubyNode string) {
+            return ToRopeNodeGen.create(string);
+        }
 
-            final int total = stringRope.byteLength();
-            int p = 0;
-            final int e = p + total;
+        @CreateCast("pattern")
+        protected RubyNode coercePatternToRope(RubyNode pattern) {
+            return ToRopeNodeGen.create(pattern);
+        }
+
+        @Specialization(guards = "offset < 0")
+        protected Object stringCharacterIndexNegativeOffset(Rope stringRope, Rope patternRope, int offset) {
+            return nil;
+        }
+
+        @Specialization(
+                guards = {
+                        "offset >= 0",
+                        "singleByteOptimizableNode.execute(stringRope)",
+                        "patternRope.byteLength() > stringRope.byteLength()" })
+        protected Object stringCharacterIndexPatternTooLarge(Rope stringRope, Rope patternRope, int offset) {
+            return nil;
+        }
+
+        @Specialization(
+                guards = {
+                        "offset >= 0",
+                        "singleByteOptimizableNode.execute(stringRope)",
+                        "patternRope.byteLength() <= stringRope.byteLength()" })
```

Comment:
I think we could check `offset + patternRope.byteLength() <=/> stringRope.byteLength()`, same below.
That way we don't try to match e.g. a pattern of 4 bytes in a string of 6 bytes with an offset of 3 bytes.
cc @nirvdrum 

I'm not sure if the offset is a number of bytes or characters in this node, but it doesn't matter for `singleByteOptimizableNode.execute(stringRope)` cases.

Reply:
In this case, it's characters. TY for bringing it up, it helped me catch a bug that specs didn't. 

I'll add the "offset + patternRope.byteLength() <=/> stringRope.byteLength()" to the singleByteOptimizable cases, I didn't check length in the guards for the other case.

Reply:
Guards don't recognize "+" or  "Integer.sum"; should I make a new function to do the addition?

Reply:
I think you'll have to yes - I'm never sure how far the subset of Java available in the guard goes!

Reply:
Integer.sum would work like `guards = "sum(a, b) < ..."` with `@ImportStatic(Integer.class)`, but it's not so readable.
So I think a method is best here for clarity (maybe `canMatch()`), and then the opposite guard can simply be `!method`.